Update setup instructions and emitter

suzannehamilton opened the pull request:
A few typo fixes and suggestions for the trainer setup.

The only code change is to the queue reaping - see comment inline.

suzannehamilton committed:
Update instructions for creating AWS EC2 key pair

suzannehamilton committed:
Fix reference to part2 in emitter setup

suzannehamilton committed:
Add instructions for setting AWS region

The region must be set before creating any resources, such as key pairs.
It also has to match the region of the EC2 image ID, or cloud formation
will fail because it can't find the t2.micro image.

suzannehamilton committed:
Do not reap pending subscriptions

suzannehamilton committed:
Fix typo in comment

suzannehamilton committed:
Add instructions for creating students' access keys

suzannehamilton commented on emitter/src/main/java/org/softwire/training/sensoremitter/consumer/SnsTopic.java:
```diff
@@ -122,6 +123,8 @@ private void processSubscription(Subscription subscription) {
                 Optional<String> queueUrlO = extractSqsQueueName(subscription.getEndpoint()).flatMap(this::getQueueUrl);
                 if (queueUrlO.isPresent()) {
                     processQueue(subscription, queueUrlO.get());
+                } else if (subscription.getSubscriptionArn().equals(PENDING_CONFIRMATION)) {
```
When the stack is created, it creates some [example queues](https://github.com/CorndelWithSoftwire/EventProcessing-Java/blob/Part2/emitter/infrastructure/cloudformation-template.json#L258) (which pick up messages from SNS correctly), and some [other queues called Q1](https://github.com/CorndelWithSoftwire/EventProcessing-Java/blob/Part2/emitter/infrastructure/cloudformation-template.json#L223) which seem to be stuck in the pending state.

This causes the emitter to crash when it first runs the subscription reaper, because the `sns.unsubscribe(subscription.getSubscriptionArn())` line below passes "PendingConfirmation" instead of an ARN.

I think this fix is useful whether or not we get rid of the Q1 queues (or get them into a non-pending state), because we don't want to accidentally delete a student's queue in the brief period where it's pending. Should we also remove them from the CloudFormation template, though?

Joe-Edwards replied:
If the example queues aren't useful then I think we should just remove them 👍 

(Unless @SimonStJG can remember if there was some other intended use?)

SimonStJG replied:
Oops sorry, I think I added the example and the Q1 queues for debugging, I didn't mean to leave them in there.

suzannehamilton replied:
I've deleted the Q1 queues from the CloudFormation template and checked that the stack can still be created. 👍

Joe-Edwards commented:
Thanks for this!

SimonStJG commented on emitter/README.md:
```diff
@@ -60,6 +60,15 @@ Go to `AWS` -> `Services` -> `SQS` -> `<Example Queue Name>`
+For each student, click on their name in the resources list and go the `Security
+credentials` tab.
+
+Click `Create access key` and copy the keys or download the CSV.
```
Just FYI, I think you can do this in cloudformation - https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-iam-accesskey.html
You could then declare the access key and secret key as cloudformation outputs.

suzannehamilton replied:
It sounds like it would generate a new key every time the stack is updated:

> When you update your stack, AWS CloudFormation will replace the existing access key with a new key.

Which might be a pain, since you'd have to give the new keys out to the students each time you update (which shouldn't happen too often, but does happen occasionally as we found when you covered for me last week).

suzannehamilton committed:
Delete trailing quote in analyser README

suzannehamilton committed:
Delete temporary SQS queues

Remove the "Q1" queues from the emitter CloudFormation template, since
they were just added for debugging purposes.

They are also permanently stuck in the "pending" state, which means they
can't be deleted by the emitter's queue reaper.

suzannehamilton commented:
Hello! I just noticed this old PR was still open.

I'm going to close it to remove it from my PR list, but feel free to reopen and merge it if it's still useful.

Joe-Edwards commented:
Thanks Suzanne - hope you're well!
